crates/transpiler/src/passes/commutative_optimization.rs

```diff
@@ -0,0 +1,516 @@
+                true,
+            ),
+            get_matrix(
+                &view2,
+                inst2.params_view(),
+                Some(matrix_max_num_qubits),
+                true,
+            ),
+        ) {
+            let product_mat = matrix1.dot(&matrix2);
+            let dim = product_mat.shape()[0] as f64;
+            let tr_over_dim = product_mat.diag().iter().sum::<Complex64>() / dim;
+
+            let (can_be_removed, phase_update) = can_remove(tr_over_dim, dim, tol);
+            if can_be_removed {
+                return Ok((true, None, phase_update));
+            }
+        }
+    }
+
+    // Could not merge.
+    Ok((false, None, 0.))
+}
+
+#[pyfunction]
+#[pyo3(name = "commutative_optimization")]
+#[pyo3(signature = (dag, commutation_checker, approximation_degree=1., matrix_max_num_qubits=0))]
+pub fn run_commutative_optimization(
+    dag: &mut DAGCircuit,
+    commutation_checker: &mut CommutationChecker,
+    approximation_degree: f64,
+    matrix_max_num_qubits: u32,
+) -> PyResult<Option<DAGCircuit>> {
+    let tol = 1e-12_f64.max(1. - approximation_degree);
+
+    let error_cutoff_fn = |_inst: &PackedInstruction| -> f64 { tol };
+
+    // Create output DAG.
+    // We will use it to intern qubits of canonicalized instructions.
+    // (In theory, we could also change qubits when merging instructions, however
+    // this does not happen right now).
+    let mut new_dag = dag.copy_empty_like(VarsMode::Alike)?;
+
+    let node_indices = dag.topological_op_nodes()?.collect::<Vec<_>>();
+    let num_nodes = node_indices.len();
+
+    let mut node_actions: Vec<NodeAction> = vec![NodeAction::Keep; num_nodes];
+    let mut new_global_phase = dag.global_phase().clone();
+
+    let mut modified: bool = false;
+
+    for idx1 in 0..num_nodes {
+        let node_index1 = node_indices[idx1];
+        let instr1 = dag[node_index1].unwrap_operation();
+
+        let (can_be_removed, phase_update) =
+            is_identity_equiv(instr1, Some(matrix_max_num_qubits), true, error_cutoff_fn)?;
+        if can_be_removed {
+            node_actions[idx1] = NodeAction::Drop;
+            new_global_phase = radd_param(new_global_phase, Param::Float(phase_update));
+            modified = true;
+            continue;
+        }
+
+        if let Some((new_instruction, phase_update)) = canonicalize(&mut new_dag, instr1) {
+            node_actions[idx1] = NodeAction::Canonical(new_instruction, phase_update);
+        }
+
+        let (instr1, extraphase1) = match &node_actions[idx1] {
+            NodeAction::Replace(instruction) => (instruction, Param::Float(0.)),
+            NodeAction::Keep => (instr1, Param::Float(0.)),
+            NodeAction::Canonical(instruction, phase) => (instruction, phase.clone()),
+            NodeAction::Drop => {
+                unreachable!("The current instruction should not be deleted.")
+            }
+        };
+
+        for idx2 in (0..idx1).rev() {
```

Comment:
Summarizing an offline discussion; there's optimization potential by only checking indices of operation on the same wire (the "wire ancestors"). This might be an optimization for later on though, we'll define a set of benchmarks to check the performance on. For now it seems that the code is on par with `CommutativeCancellation`, but produces a better quality result.

Reply:
I definitely agree with this plan. I can't immediately think of simple solutions that also do not require an excessive amount of bookkeeping, so let's leave this for the follow-up.